### PR TITLE
MESH-2060/Add more checks to execute_instruction

### PR DESCRIPTION
**Pull request by HenriqueNogara**
Adds multiple checks for execute_instruction

### CODE REVIEW

File: pallets/settlement/src/lib.rs [hunk 1]
```diff
@@ -1295,13 +1307,75 @@ impl<T: Config> Module<T> {
...
+                        );
+                        let rcv_affirmation_status =
+                            AffirmsReceived::take(instruction_id, receiver);
+                        ensure!(
+                            rcv_affirmation_status == AffirmationStatus::Affirmed,
+                            Error::<T>::NotAllAffirmationsHaveBeenReceived
+                        );
+                    }
+                }
+                Leg::OffChain { .. } => continue,
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Could take and check the offchain receipt status here.

File: pallets/settlement/src/lib.rs [hunk 2]
```diff
@@ -1343,56 +1417,52 @@ impl<T: Config> Module<T> {
...
 
-        for (_, leg) in drained_legs {
+    /// Clears the storage for a rejected instruction and updates the instruction status to
+    /// [`InstructionStatus::Rejected`].
+    fn prune_rejected_instruction(instruction_id: InstructionId) {
+        let instruction_details = InstructionDetails::<T>::take(&instruction_id);
+        VenueInstructions::remove(instruction_details.venue_id, instruction_id);
+        InstructionAffirmsPending::remove(instruction_id);
+        let _ = OffChainAffirmations::clear_prefix(instruction_id, u32::MAX, None);
+        let _ = AffirmsReceived::clear_prefix(instruction_id, u32::MAX, None);
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: These two can be cleaned up in the for loop below.

File: pallets/settlement/src/lib.rs [hunk 3]
```diff
@@ -1343,56 +1417,52 @@ impl<T: Config> Module<T> {
...
+        let _ = InstructionMediatorsAffirmations::<T>::clear_prefix(instruction_id, u32::MAX, None);
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]:
```suggestion
        let _ = InstructionMediatorsAffirmations::<T>::clear_prefix(instruction_id, T::MaxInstructionMediators, None);
```

File: pallets/settlement/src/lib.rs [hunk 4]
```diff
@@ -1209,74 +1205,90 @@ impl<T: Config> Module<T> {
...
+                }
+                Err(leg_id) => {
+                    Self::deposit_event(RawEvent::LegFailedExecution(
+                        caller_did,
+                        instruction_id,
+                        leg_id,
+                    ));
+                    Self::deposit_event(RawEvent::InstructionFailed(caller_did, instruction_id));
+                    Err(Error::<T>::FailedToReleaseLockOrTransferAssets.into())
+                }
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: This is still inside the `with_transaction`, so it would be reverted too and those events would be lost.

**Review by Neopallium — CHANGES_REQUESTED** [review]
There is an issue with events being lost, if the settlement fails to execute.

File: pallets/settlement/src/lib.rs [hunk 5]
```diff
@@ -1295,13 +1313,69 @@ impl<T: Config> Module<T> {
...
+                        let sdr_affirmation_status = UserAffirmations::take(sender, instruction_id);
+                        ensure!(
+                            sdr_affirmation_status == AffirmationStatus::Affirmed,
+                            Error::<T>::NotAllAffirmationsHaveBeenReceived
+                        );
+                        let sdr_affirmation_status = AffirmsReceived::take(instruction_id, sender);
+                        ensure!(
+                            sdr_affirmation_status == AffirmationStatus::Affirmed,
+                            Error::<T>::NotAllAffirmationsHaveBeenReceived
+                        );
```
Comment (adamdossa) [inline pallets/settlement/src/lib.rs:1342]: I can't see any way for `UserAffirmations` and `AffirmsReceived` to be out of sync (i.e. for one to be Affirmed whilst the other is not) - are there any known ways for this to happen?

File: pallets/settlement/src/lib.rs [hunk 6]
```diff
@@ -1209,74 +1205,96 @@ impl<T: Config> Module<T> {
...
-            .check_accrue(<T as Config>::WeightInfo::execute_instruction_paused(
-                instruction_asset_count.fungible(),
-                instruction_asset_count.non_fungible(),
-                instruction_asset_count.off_chain(),
-            ))
-            .map_err(|_| Error::<T>::WeightLimitExceeded)?;
+        let mut failed_leg_id = None;
+        let tx_result = with_transaction(|| {
+            // Ensures the number of pending affirmations is zero
+            let n_pending_affirmations = InstructionAffirmsPending::take(instruction_id);
```
Comment (adamdossa) [inline pallets/settlement/src/lib.rs:1211]: Is there any known way for `InstructionAffirmsPending` to be 0, but for `Self::ensure_no_missing_affirmation` to fail?

Comment (Neopallium) [reply pallets/settlement/src/lib.rs:1211]: It shouldn't be possible for that to happen, unless a bug is introduced later.  All of the storage for affirmations needs to be cleared anyways, so it isn't much more expensive to check the values.

File: pallets/settlement/src/lib.rs (hunk 5 above)
Comment (Neopallium) [reply pallets/settlement/src/lib.rs:1342]: It shouldn't, but we need to remove the old values anyways.  Those two values are simple/small, so it is cheap to decode and check the value.  We could just use `remove` for one and still use `take` for the other.

**Review by Neopallium — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Pull request merged**